k2/python/csrc/torch/index_select.cu

```diff
@@ -0,0 +1,143 @@
+/**
+ * @brief Index select for k2.
+ *
+ * Unlike torch.index_select, when an entry is -1, it sets
+ * the destination entry to 0.
+ *
+ * @copyright
+ * Copyright (c)  2020  Mobvoi Inc.        (authors: Fangjun Kuang)
+ *
+ * @copyright
+ * See LICENSE for clarification regarding multiple authors
+ */
+
+#include "k2/csrc/context.h"
+#include "k2/python/csrc/torch/index_select.h"
+#include "k2/python/csrc/torch/torch_util.h"
+#include "torch/extension.h"
+
+namespace k2 {
+
+template <typename T>
+static torch::Tensor IndexSelect1D(torch::Tensor src, torch::Tensor index) {
+  K2_CHECK_EQ(src.dim(), 1) << "Expected dim: 1. Given: " << src.dim();
+  K2_CHECK_EQ(src.scalar_type(), ToScalarType<T>::value);
+  K2_CHECK(src.is_contiguous());
+
+  K2_CHECK(index.is_contiguous());
+  K2_CHECK_EQ(index.dim(), 1);
+  K2_CHECK_EQ(index.scalar_type(), ToScalarType<int32_t>::value);
+
+  ContextPtr context;
+  if (src.device().type() == torch::kCPU) {
+    context = GetCpuContext();
+  } else {
+    K2_CHECK(src.is_cuda());
+    context = GetCudaContext(src.device().index());
+  }
+
+  const T *src_data = src.data_ptr<T>();
+  int32_t src_numel = src.numel();
+  const int32_t *index_data = index.data_ptr<int32_t>();
+
+  torch::Tensor ans = torch::empty(index.sizes(), src.options());
+  T *ans_data = ans.data_ptr<T>();
+
+  auto lambda = [=] __host__ __device__(int32_t i) {
```

Comment:
I tend to prefer to implement the CPU version with a loop rather than a lambda for things like this, because the current implementation of lambdas from nvcc has to follow a function pointer each time the lambda is invoked, i.e. it's not inline.  So you could do:
```
   if (context->GetDeviceType() == kCpu) {
        // use a loop
   } else {
       EvalDevice(context, index.numel(), lambda);
  }
```
I introduced EvalDevice() fairly recently so you won't see it in a ton of code yet.  It just means, we won't call this for CPU.
```

Reply:
Fixed.